<div style="border:solid purple 2px; padding: 20px"> 

Александра, привет! 👋

Меня зовут Рустам Муртазин, и я буду делать ревью проектной работы по предобработке данных (точнее ее части). Давай будем общаться на **«ты»**. Если это неприемлемо, обязательно напиши мне в комментариях — мы перейдем на **«вы»**.

В любом деле всегда можно можно что-то улучшить. Моя основная цель — не показать, что где-то совершены какие-либо ошибки, а поделиться своим опытом, который поможет тебе в дальнейших работах. Далее в файле ты сможешь увидеть мои комментарии, постарайся учесть их при выполнении следующих заданий.

Обрати внимание в первую очередь на те, что помечаны <span style="color:red">красным цветом</span>. После их доработки проект будет принят. <span style="color:green">Зеленым цветом</span> отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. <span style="color:orange">Оранжевым цветом</span> выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно). Также в проекте могут быть небольшие «лайфхаки» по Python, не относящиеся к проекту, их я выделил в фиолетовую рамку)

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее».

Мне понравился твой проект, есть пару замечаний (одно, если быть точным), я думаю ты быстро справишься. Жду твою работу на повторное ревью :)

![](https://i.gifer.com/7V3.gif)

<div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 ✔️ <a class="tocSkip"> </h2>

Показал пример реализации с помощью pivot_table, думаю, он тебе поможет... Надеюсь ты учтешь мои комментарии в последующих работах. Благодарю за старания и интерес и желаю успехов в дальнейших проектах 😊

</div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

54

In [16]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [17]:
data['education'] = data['education'].str.lower()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

In [23]:
data.groupby('children')['debt'].sum() # количество задолжностей

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

In [24]:
data.groupby('children')['children'].count() # количество людей с разных количеством детей

children
0    14107
1     4809
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

In [25]:
data.groupby('children')['debt'].sum() / data.groupby('children')['children'].count() *100 # доля задолжностей, отнесенная 
# к количеству людей с данным количеством детей

children
0    7.535266
1    9.232689
2    9.454191
3    8.181818
4    9.756098
5    0.000000
dtype: float64

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

А вот как можно было сократить решение. Обрати внимание на [style](https://pandas.pydata.org/pandas-docs/stable/user_guide/style.html) библиотеки `pandas`. Там много прикольных штук для того, чтобы сделать датафрейм красивее 😊

In [26]:
data['children_group'] = data['children'].apply(lambda x: 'Нет детей' if x == 0 else 'Дети есть')

df_example = data.groupby('children_group')['debt'].agg(['count', 'sum', 'mean', lambda x: 1 - x.mean()])
df_example.columns = ['Кол-во пользователей', 'Кол-во должников', '% должников', '% НЕдолжников']
df_example.style.format({'% должников': '{:.2%}', '% НЕдолжников': '{:.2%}'})

**Вывод:** Явной зависимости между количеством детей и возвратом кредита в срок есть. Доля задолжностей у людей количеством детей 1,2 и 4 вырьируется от 9,2 до 9,8%. Доля задолжностей у людей с 3 детьми и  без детей составляет 8,18 и 7,5%, соотвественно. Делать выводы по поводу категории людей с пятью детьми трудно, т.к. выборка состоит из 9 случаев. 

Получается, наименьшая вероятность задолжности кредита у людей без детей.

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [27]:
data.groupby('family_status')['debt'].sum()
# print (debts)

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

In [28]:
debts = data.pivot_table(index='family_status', values='debt', aggfunc='sum')
print(debts)

                       debt
family_status              
Не женат / не замужем   273
в разводе                84
вдовец / вдова           63
гражданский брак        385
женат / замужем         927


In [29]:
data.groupby('family_status')['family_status'].count()


family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4146
женат / замужем          12266
Name: family_status, dtype: int64

In [30]:
pd.concat((debts,data.groupby('family_status')['family_status'].count(),data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['family_status'].count() * 100),axis=1)




debt  family_status         0
family_status                                       
Не женат / не замужем   273           2796  9.763948
в разводе                84           1189  7.064760
вдовец / вдова           63            951  6.624606
гражданский брак        385           4146  9.286059
женат / замужем         927          12266  7.557476

**Вывод:** зависимость между семейным положением и возвратом кредита в срок существует. Категории "Не женат / не замужем" и "гражданский брак" имеют самый большой процент задолжностей - 9,76% и 9,29%, соотвественно. Затем следуют категории "женат / замужем  " и "в разводе", имеющие долю задолжностей 7,56% и 7,06%, соотвественно. Наименьшую долю задолжностей имеет категория "вдовец / вдова" -  6,62%. 

<div class="alert alert-danger">
<h2> Комментарий ревьюера ❌ <a class="tocSkip"> </h2>

Альтернатива `groupby` - это [pivot_table](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html). Это очень популярный и нужный инструмент, важно иметь его в арсенале, и мне нужно убедиться, что ты умеешь им пользоваться. Реализуй то же самое используя `pivot_table()` и дай комментарии, какой из методов удобнее для тебя.

КОММЕТАРИЙ:

использовала pivot_table в первом случае при нахождениеи количества debts. В остальных случаях выдавал ошибку,например, "Grouper for 'family_status' not 1-dimensional", поэтому просто решила соединить эти данные в одну таблицу. В одной функции pivot_table тоже не получилось сделать. 

groupby удобнее метод, тк мы больше на нем тренировались, и понятно, что от чего следует. pivot_table по сути изучала сама. 

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 ✔️ <a class="tocSkip"> </h2>

Было бы круто, если ты оставила код, тогда я бы точечно мог тебе подсказать, в чем проблема... Показал, как то же самое можно реализовать с помощью pivot_table

In [31]:
df_example = data.pivot_table(index = 'children_group', values = 'debt', 
                            aggfunc = ['count', 'sum', 'mean', lambda x: 1 - x.mean()])
df_example.columns = ['Кол-во пользователей', 'Кол-во должников', '% должников', '% НЕдолжников']
df_example.style.format({'% должников': '{:.2%}', '% НЕдолжников': '{:.2%}'})

<div class="alert alert-info">
<h2> Комментарий ревьюера v2 <a class="tocSkip"> </h2>

И еще, свои комментария лучше выделять цветом, можно так) Оставлю еще [ссылку на Medium](https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd), там очень круто расписаны Markdown, и еще вот такую [шпаргалку](https://paulradzkov.com/2014/markdown_cheatsheet/), вдруг понадобится 🙂
    
</div>

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [32]:
data.groupby('total_income_category')['debt'].sum() # Ваш код будет здесь. Вы можете создавать новые ячейки.

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

In [33]:
print(data.groupby('total_income_category')['total_income_category'].count())
data.info()

total_income_category
A       25
B     5014
C    15938
D      349
E       22
Name: total_income_category, dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 21348 entries, 0 to 21524
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21348 non-null  int64  
 1   days_employed          21348 non-null  float64
 2   dob_years              21348 non-null  int64  
 3   education              21348 non-null  object 
 4   education_id           21348 non-null  int64  
 5   family_status          21348 non-null  object 
 6   family_status_id       21348 non-null  int64  
 7   gender                 21348 non-null  object 
 8   income_type            21348 non-null  object 
 9   debt                   21348 non-null  int64  
 10  total_income           21348 non-null  int64  
 11  purpose                21348 non-null  object 
 12  total_income_category  21348 non-null  

In [34]:
data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['total_income_category'].count() * 100

total_income_category
A    8.000000
B    7.060231
C    8.489145
D    6.017192
E    9.090909
dtype: float64

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Здесь, наверное, стоит подумать над другой разбивкой, например, по процентилям. Слишком большая разница в кол-ве пользователей по группам. Тогда можно воспользоваться методами [cut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html) или [qcut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html)(в этом случае деление будет проводиться по проценилям). Обрати внимание, у данного метода есть параметр `labels`, так можно оформить группы красивее 😊

**Вывод:** зависимость между уровнем дохода и возвратом кредита в срок существует. Категория "С" имеет один из наибольших процентов задолжностей - 8,49%. Затем следуют категории "B" и "D", имеющие долю задолжностей 7,06% и 6,02%, соотвественно. Говорить о том, что категория "E" имеет самый большой процент задолжностей трудно, потому что количество людей данной категории равно 22 из 21348 возможных, то есть недостаточно кейсов. Тоже самое про группу "A".

Но, в целом, процент задолжностей категорий варьируется от 6% до 9%. 

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

In [35]:
data.groupby('purpose_category')['debt'].sum()# Ваш код будет здесь. Вы можете создавать новые ячейки.

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

In [36]:
data.groupby('purpose_category')['purpose_category'].count()

purpose_category
операции с автомобилем       4281
операции с недвижимостью    10754
получение образования        3989
проведение свадьбы           2324
Name: purpose_category, dtype: int64

In [37]:
data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['purpose_category'].count() * 100

purpose_category
операции с автомобилем      9.343611
операции с недвижимостью    7.253115
получение образования       9.250439
проведение свадьбы          7.874355
dtype: float64

<div style="border:solid purple 5px; padding: 20px"> 
<h2 align="center"> Рубрика «Питонячий лайфхакер» <a class="tocSkip"> </h2>

<h3> Широкоформатный Jupyter <a class="tocSkip"> </h3>

Расширяем границы, или как сделать работу более комфортной (не всем нравится 😄)

    from IPython.core.display import display, HTML
    display(HTML("<style>.container { width:90% !important; }</style>"))

**Вывод:** Категории целей "операции с автомобилем" и "получение образования " имеют самый большой процент задолжностей - 9,34% и 9,25%, соотвественно. Затем следуют категории целей "проведение свадьбы " и "операции с недвижимостью", имеющие долю задолжностей 7,87% и 7,25%, соотвественно.

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

*Ответ:* пропуски в исходных данных были в колоннах days_employed и total_income. Т.к. количество пропусков в столбцах стажа и дохода одинаковое, есть вероятность, что это случайные пропуски. То есть без стажа работы нет и ежемесячного дохода.

Также могут быть полностью случайные пропуски, то есть случайно не указаны данные о стаже и доходе. 

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ:* медианное значение более корректно характеризует данные, когда некоторые значения сильно выделяются среди большинства, чем среднее значение. Например:

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Да, согласен

In [38]:
data_pivot = data.pivot_table(index='income_type', values='days_employed', aggfunc=['mean','median'])
print(data_pivot)

                          mean         median
                 days_employed  days_employed
income_type                                  
безработный      366413.652744  366413.652744
в декрете          3296.759962    3296.759962
госслужащий        3332.636454    2689.137274
компаньон          2062.683448    1555.947387
пенсионер        365051.861219  365286.622650
предприниматель     520.848083     520.848083
сотрудник          2250.511445    1572.328285
студент             578.751554     578.751554


Видим, что среднее и медианное значения разнятся. Медианное значение тем лучше, что при выборке, например, 100,100,100, 700, пропущенное значение будет занято 100, приближенное к общей тенденции выборки, а не 250, которое является средним арифметическим. 

## Общий вывод.

В ходе анализа было выяснено, что семейное положение, количество детей, цель и уровень дохода влияет на факт погашения кредита в срок. 

В пределах категорий процент задолжности  варьировался от 6 до 9,5%. 

С самой большой долей задолжностей оказались категории с количеством детей 1,2 и 4, "Не женат / не замужем" и "гражданский брак","операции с автомобилем" и "получение образования " - выше 9%. 

С наименьшей долей задолжностей оказались категории "без детей", "в разводе" и "вдовец / вдова", "B" и "D", "операции с недвижимостью" - менее 7,5%